kv/etcd/store_test.go

```diff
@@ -653,6 +735,85 @@ func TestDelete_TriggerWatch(t *testing.T) {
 	verifyValue(t, vw.Get(), "bar3", 1)
 }
 
+func TestStaleDelete(t *testing.T) {
+	// in this test we ensure clients who did not receive a delete for a key in
+	// their caches, evict the value in their cache the next time they communicate
+	// with an etcd which is unaware of the key (e.g. it's been compacted).
+
+	// first, we find the bytes required to be created in the cache file
+	serverCachePath, err := ioutil.TempDir("", "server-cache-dir")
+	require.NoError(t, err)
+	defer os.RemoveAll(serverCachePath)
+	ec, opts, closeFn := testStore(t)
+
+	setStore, err := NewStore(ec, ec, opts.SetCacheFileFn(func(ns string) string {
+		return path.Join(serverCachePath, fmt.Sprintf("%s.json", ns))
+	}))
+	require.NoError(t, err)
+
+	setClient := setStore.(*client)
+	version, err := setClient.Set("foo", genProto("bar1"))
+	require.NoError(t, err)
+	require.Equal(t, 1, version)
+
+	setV, err := setClient.Get("foo")
+	require.NoError(t, err)
+	verifyValue(t, setV, "bar1", 1)
+	require.Equal(t, 1, len(setClient.cache.Values))
+
+	// drain the notification to ensure set received update
+	var (
+		fileName   string
+		cacheBytes []byte
+	)
+	require.True(t, xclock.WaitUntil(func() bool {
+		fileName, cacheBytes, err = readCacheJSONAndFilename(serverCachePath)
+		return err == nil
+	}, 2*time.Second), "timed out waiting to read cache file")
+	closeFn()
+
+	// make a new etcd cluster (to mimic the case where etcd has deleted and compacted
+	// the value).
+	newServerCachePath, err := ioutil.TempDir("", "new-server-cache-dir")
+	require.NoError(t, err)
+	defer os.RemoveAll(newServerCachePath)
+
+	// write the cache file with correct contents in the new location
+	f, err := os.Create(path.Join(newServerCachePath, fileName))
+	require.NoError(t, err)
+	_, err = f.Write(cacheBytes)
+	require.NoError(t, err)
+	require.NoError(t, f.Sync())
+	require.NoError(t, f.Close())
+
+	require.True(t, xclock.WaitUntil(func() bool {
+		_, newBytes, err := readCacheJSONAndFilename(newServerCachePath)
+		return err == nil && bytes.Equal(cacheBytes, newBytes)
+	}, 2*time.Second), "timed out waiting to flush new cache file")
+
+	// create new etcd cluster
+	ec2, opts, closeFn2 := testStore(t)
+	closeFn2()
+	// defer closeFn()
+	getStore, err := NewStore(ec2, ec2, opts.SetCacheFileFn(func(ns string) string {
+		nsFile := path.Join(newServerCachePath, fmt.Sprintf("%s.json", ns))
+		return nsFile
+	}))
+	require.NoError(t, err)
+	getClient := getStore.(*client)
+
+	require.True(t, xclock.WaitUntil(func() bool {
+		return 1 == len(getClient.cache.Values)
+	}, 2*time.Second), "timed out waiting for client to read values from cache")
+	println(fmt.Sprintf("%+v", getClient.cache.Values))
+
+	// get value and ensure it's able to serve the read
+	v, err := getClient.Get("foo")
+	require.NoError(t, kv.ErrNotFound)
+	require.NotNil(t, v)
+	// TODO: should this not serve the read? it's got the value in the cache and can't talk to etcd.
```

Comment:
while trying to repro matt's case i ended up here: it appears that this isn't serving a read from the cache while etcd is down. 

i'm probably missing something atm but reminder for us to talk about it @cw9